tests/src/App/DocumentObserver.cpp

```diff
@@ -0,0 +1,426 @@
+    // An App::SubObjectT object used to test SubObjectT::normalize() with the option argument set
+    // to SubObjectT::NormalizeOption::ConvertIndex
+    auto subObjTConvertIndex {SubObjectT(lGrp, subName.c_str())};
+
+    // A bool variable used to store the result of subObjTEmpty.normalize()
+    auto normalizeEmpty {false};
+    // A bool variable used to store the result of subObjTWithoutSubObj.normalize()
+    auto normalizeWithoutSubObj {false};
+    // A bool variable used to store the result of subObjTWithSubObj.normalize()
+    auto normalizeWithSubObj {false};
+    // A bool variable used to store the result of subObjTWithoutEl.normalize()
+    auto normalizeWithoutEl {false};
+    // A bool variable used to store the result of subObjTNoFlatten.normalize()
+    auto normalizeNoFlatten {false};
+    // A bool variable used to store the result of subObjTKeepSubName.normalize()
+    auto normalizeKeepSubName {false};
+    // A bool variable used to store the result of subObjTConvertIndex.normalize()
+    auto normalizeConvertIndex {false};
+
+    // Act
+
+    normalizeEmpty = subObjTEmpty.normalize();
+    normalizeWithoutSubObj = subObjTWithoutSubObj.normalize();
+    normalizeWithSubObj = subObjTWithSubObj.normalize();
+    normalizeWithoutEl = subObjTWithoutEl.normalize(SubObjectT::NormalizeOption::NoElement);
+    normalizeNoFlatten = subObjTNoFlatten.normalize(SubObjectT::NormalizeOption::NoFlatten);
+    normalizeKeepSubName = subObjTKeepSubName.normalize(SubObjectT::NormalizeOption::KeepSubName);
+    normalizeConvertIndex =
+        subObjTConvertIndex.normalize(SubObjectT::NormalizeOption::ConvertIndex);
+
+    // Assert
+
+    // In this case calling SubObjectT::normalize() doesn't have effect as subObjTEmpty has been
+    // initialized with an empty constructor
+    EXPECT_FALSE(normalizeEmpty);
+
+    // In this case calling SubObjectT::normalize() doesn't have effect as subObjTWithoutSubObj
+    // hasn't any sub objects
+    EXPECT_FALSE(normalizeWithoutSubObj);
+
+    // In this case calling SubObjectT::normalize() changes subObjTWithSubObj subname removing the
+    // "Fusion." part of the of the original subname
+    EXPECT_TRUE(normalizeWithSubObj);
+
+    // In this case calling SubObjectT::normalize() changes subObjTWithoutEl subname removing the
+    // "Fusion." and "Edge1" parts of the of the original subname
+    EXPECT_TRUE(normalizeWithoutEl);
+
+    // In this case calling SubObjectT::normalize() doesn't have effect as neither the
+    // DocumentObject referenced nor the subname of subObjTNoFlatten are changed
+    EXPECT_FALSE(normalizeNoFlatten);
+
+    // In this case calling SubObjectT::normalize() changes subObjTKeepSubName subname removing the
+    // "Fusion." part of the of the original subname
+    EXPECT_TRUE(normalizeKeepSubName);
+
+    // In this case calling SubObjectT::normalize() changes subObjTConvertIndex subname replacing
```

Comment:
It seems as if `normalizeKeepSubName` has the "0" in the subname, so there is probably a mismatch in the tests. Could that be the case? 

In any case, if I understand correctly, normalizing with converting indices allows you to use an index in the subname that then is replaced by a the part name? Does this mean that subnames may have indices inside the string that sometimes have a special meaning? What if the subname has something that looks like an index in the name, for example "Edge1" that was used above?

Reply:
This test was a little bit misleading because of a misunderstanding by my side of what the logic does.

The idea was to have both `subObjTKeepSubName` and `subObjTConvertIndex` with the "0" in the subname and see that `subObjTKeepSubName` keeps it while `subObjTConvertIndex` converts it back to "App__Part" (which is the value of `partName`).

However the code does something different and therefor I reworked some of the tests in my last commit.

Briefly, there are cases where `DocumentObject`s can be selected using indexes, and in those cases `NormalizeOption::ConvertIndex` converts the indexes into the names (not the Labels) of the related`DocumentObject`s.

`NormalizeOption::KeepSubName` instead is used, for example, to convert or not the Label into the name.
If I'm not wrong there are other cases where it does conversions but in my last commit I tested only the Label<->name conversion.